components/script/textinput.rs

```diff
@@ -393,6 +393,148 @@ impl<T: ClipboardProvider> TextInput<T> {
+    fn word_start_adjustment(&self) -> isize {
+        let mut adjust = 0;
+        let mut found_non_whitespace = false;
+        let mut edit_point_index = Some(self.edit_point.index);
+        let mut line_iter = self.lines[..(self.edit_point.line + 1)].iter();
+
+        'line: loop {
+            match line_iter.next_back() {
+                Some(line) => {
+                    let mut char_iter = match edit_point_index {
+                        Some(i) => line[..i].chars(),
+                        None => line.chars(),
+                    };
+
+                    'chars: loop {
+                        match char_iter.next_back() {
+                            Some(ch) => {
+                                let is_whitespace = ch.is_whitespace();
+
+                                // We want to ignore whitespace until we enter a word, then stop at
+                                // either the first whitespace we find or the end of the text.
+                                if is_whitespace && found_non_whitespace { break 'line; }
+                                // Ensure we record if we found a non whitespace char.  We may not,
+                                // in which case we have to jump to the next line.
+                                if !is_whitespace { found_non_whitespace = true; }
+
+                                adjust = adjust - (ch.len_utf8() as isize);
+                            },
+                            None => break 'chars,
+                        }
+                    }
+
+                    // If we've found some part of a word, we can stop at the end of the
+                    // current line.  Otherwise, we need to jump to the end of the first word
+                    // of the next line.
+                    if found_non_whitespace { break 'line; }
+
+                    adjust = adjust - 1; // Account for newline character
+                    edit_point_index = None;
+                },
+                None => break 'line,
+            }
+        }
+
+        adjust
+    }
+
+    /// Returns the adjustment needed to move to the end of the next or current word.
+    ///
+    /// "he<cursor>llo world\n\nhi" would be 3, since the cursor must move to the end of "hello".
+    /// "hello<cursor> world\n\nhi" would be 6, since the cursor must move to the end of "world".
+    /// "hello world<cursor>\n\nhi" would be 4, since the cursor must move to the end of "hi".
+    fn word_end_adjustment(&self) -> isize {
+        let mut adjust = 0;
+        let mut found_non_whitespace = false;
+        let mut edit_point_index = Some(self.edit_point.index);
+
+        'line: for line in self.lines[self.edit_point.line..].iter() {
+            let char_iter = match edit_point_index {
+                Some(i) => line[i..].chars(),
+                None => line.chars(),
+            };
+
+            'chars: for ch in char_iter {
```

Comment:
Could just be `for ch in line[edit_point_index.unwrap_or(0)..].chars() {`. Also no need for the `'chars:`